### PR TITLE
Rage generation on received damage

### PR DESCRIPTION
**Pull request by geo-tp**
Formula :  Rage Gained = (Damage Taken) / (Rage Conversion at Your Level) * 2.5 

Generate rage on received damage when units have def/battle stance or bear form
(Berserker stance does not generate rage in 0.5.3)

Move rage logic from PlayerMgr to UnitMgr so all units can use rage

### CODE REVIEW

File: utils/Formulas.py [hunk 1]
```diff
@@ -74,6 +74,17 @@ def get_gray_level(player_level):
 
         return gray_level
 
+    @staticmethod
+    def calculate_rage_regen_on_received_damage(damage_info):
+        # Vanilla rage formula source http://blue.mmo-champion.com/topic/18325-the-new-rage-formula-by-kalgan/
+        rage_gained = (damage_info.damage / PlayerFormulas.rage_conversion_value(damage_info.victim.level)) * 2.5
+
+        # 2458 Berserker stance - generate 30% more rage
+        if (damage_info.victim.aura_manager.get_auras_by_spell_id(2458)):
```
Comment (Fluglow) [inline utils/Formulas.py:83]: No need to check for an ID here, use has_form with the berserker stance form instead.

File: game/world/managers/objects/units/player/PlayerManager.py [hunk 2]
```diff
@@ -1241,6 +1241,13 @@ def generate_rage(self, damage_info, is_player=False):
                                                            self.has_form(ShapeshiftForms.SHAPESHIFT_FORM_BEAR)):
             self.set_rage(self.power_2 + Formulas.PlayerFormulas.calculate_rage_regen(damage_info, is_player=is_player))
 
+    def generate_rage_on_received_damage(self, damage_info):
+        # Warriors or Druids in Bear form
+        if self.player.class_ == Classes.CLASS_WARRIOR or (self.player.class_ == Classes.CLASS_DRUID and
```
Comment (GrenderG) [inline game/world/managers/objects/units/player/PlayerManager.py:1246]: Warriors can only regen rage when they receive damage if they are in Defensive or Battle stances.

File: utils/Formulas.py (hunk 1 above)
Comment (geo-tp) [reply utils/Formulas.py]: it's done, thanks

File: game/world/managers/objects/units/player/PlayerManager.py [hunk 3]
```diff
@@ -1241,6 +1241,13 @@ def generate_rage(self, damage_info, is_player=False):
...
+                                                           self.has_form(ShapeshiftForms.SHAPESHIFT_FORM_BEAR)):
```
Comment (Fluglow) [inline game/world/managers/objects/units/player/PlayerManager.py:1247]: No need to check for class. By checking for just shapeshift forms, the method is also compatible with creatures.

File: game/world/managers/objects/units/player/PlayerManager.py (hunk 2 above)
Comment (geo-tp) [reply game/world/managers/objects/units/player/PlayerManager.py]: it's done

File: game/world/managers/objects/units/player/PlayerManager.py (hunk 3 above)
Comment (geo-tp) [reply game/world/managers/objects/units/player/PlayerManager.py]: it's done

**Pull request merged**